### ISSUE TITLE
POST issues

### ISSUE THREAD

**Issue by Piero01** [warmcat/libwebsockets#2166]
Hello,

I have some problems with the latest versions to receive post requests on server with http2.

I tested with the test-server example (both local and https://libwebsockets.org/testserver/).
If I try to send form 1 to the server with Edge, the formtest page is never showed. It seems Edge is retrying for 1 minute without getting success or error message from the server.
If I list the callbacks in the callback_post_demo() function in plugin I get:

> 
> [2021/01/18 10:38:19:8160] U: 	POST callback reason: 71
> [2021/01/18 10:38:19:8198] U: 	POST callback reason: 49
> [2021/01/18 10:38:19:8199] U: 	POST callback reason: 12
> [2021/01/18 10:38:19:8200] U: 	POST callback reason: 13
> [2021/01/18 10:38:19:8201] U: 	POST callback reason: 50
> [2021/01/18 10:38:19:8202] U: 	POST callback reason: 5

So after LWS_CALLBACK_HTTP_BODY (13) we should have LWS_CALLBACK_HTTP_BODY_COMPLETION (14) but we are dropping protocol instead.

If it can help it's the same with Chrome and Firefox when the method is not sent in the request. It happens sometimes but I don't know how to reproduce this behavior. Eg sometimes it's sending the request like:
`POST /foo undefined`
instead of 
`POST /foo HTTP/2`

One other thing related to POST, the post URI (content of variable 'in' in LWS_CALLBACK_HTTP) is different if the request is HTTP2 or not.
With a mount point "/foo" (with .origin_protocol = LWSMPRO_CALLBACK) and request to https://x.x.x.x/foo:
- http1.1: in = ""
- http2: in = "/foo"

shouldn't we have the same whatever the method as it was before?

Thanks.

**Comment by lws-team** [comment]
> POST /foo HTTP/2

H2 doesn't actually look like that, it doesn't use h1 style headers but HPACK encoding.

I think I could reproduce the problem, what I found is caused by a race as to whether the body is coalesced with the header part or not... there's a fix on current main and also libwebsockets.org is updated with it too.

https://libwebsockets.org/git/libwebsockets/commit/lib/event-libs?id=aa7d852359e3c259f1a4d4acee74ce8b47763c9a

there's another patch in there as well that should correct the h2 post _HTTP in + len.

https://libwebsockets.org/git/libwebsockets/commit/lib/event-libs?id=a90fa2b776563319f746f1c890b207971dd8af33

**Comment by Piero01** [comment]
Thanks for the patches. The second one seems to be good but unfortunately the first one did not solve the issue.
I just tried with branch origin/v4.1-stable and it was working fine.

I'm not used to H2. What I noticed is the difference I can see in Firefox Network Development tool between a working request and a not working one. Maybe it's not the cause of the problem.

I don't know how I can help. I enabled info logging in lws and here is the log of a request sent with Edge:
[Edge post issue.txt](https://github.com/warmcat/libwebsockets/files/5831817/Edge.post.issue.txt)

**Comment by lws-team** [comment]
Yeah... there's a line missing after cleaning it up to push it, sorry.

I put it back and force-push the updated first patch on main, that should solve it.

**Comment by Piero01** [comment]
Thanks it seems ok now.
